src/sig.c

```diff
@@ -203,16 +206,60 @@ void do_kill(int signo)
 				else if (file[0] == '@')
 					_d("Skipping %s ...", &file[1]);
 				else
-					kill(pid, signo);
+					if (callback(pid, context)) {
+						print(0, "PID %d is still alive (%s)", pid, file);
+						break;
+					}
 			}
 			fclose(fp);
 		}
 		closedir(dirp);
 	}
 }
 
+static int kill_callback(int pid, int *context)
+{
+	kill(pid, (int)context);
+	return 0;
+}
+
+static int status_callback(int pid, int *context)
+{
+	*context = 1;
+	return 1;
+}
+
+/* return value:
+ *  1 - at least one process remaining
+ *  0 - no processes remaining
+ */
+static int do_wait(int secs)
+{
+	int has_proc;
+	int tmo = 250000;
+	int iterations = secs*1000*1000/tmo;
+
+	do {
+		do_usleep(tmo);
+		while (waitpid(-1, NULL, WNOHANG) > 0)
+			;
+		has_proc = 0;
+		iterations--;
+		do_iterate_proc(status_callback, &has_proc);
+	}
+	while (has_proc && iterations > 0);
+
+	return has_proc;
+}
+
 void do_shutdown(shutop_t op)
 {
+	struct sched_param sched_param;
+	sched_param.sched_priority = 99;
+
+	/* PID 1 run as RR process with highest priority */
+	sched_setscheduler(1, SCHED_RR, &sched_param);
```

Comment:
The comment is useless as-is.  It states in text what the next line does ... which can be read from the code itself. Instead, use comments to explain *why* a code path is needed. Here's a good place to add the reasoning from the discussion.  E.g., *"On a PREEMPT-RT system, Finit must run as the highest prioritized RT process to ensure it completes the shutdown sequence."*

Reply:
@troglobit Good point, will do that.